File: rest_api/ruby/fetch_engagements.rb
#!/usr/bin/ruby

require 'net/http'
require 'json'

# Bearer token
TOKEN = "eyJhbGciOiJIU<...truncated>"

def fetch_engagements(uri)
  res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: uri.scheme == 'https') do |http|
    # Prepare request with headers
    request = Net::HTTP::Get.new(uri)
    request['Accept'] = 'application/vnd.salemove.v1+json'
    request['Authorization'] = "Bearer #{TOKEN}"

    # Make request
    http.request(request)
  end

  # Returns response body if request is successful
  return res.body if res.is_a?(Net::HTTPSuccess)

  # Prints out the error
  puts res.body
  # Raises error to stop execution
  raise StandardError
end

all_engagements = []

# Setup first query
next_uri = URI('https://api.salemove.com/engagements')
next_uri.query = 'site_ids[]=24c9802c-3b95-4a15-b366-342b0a8289f7&per_page=100'

# While next_uri is provided
while next_uri
  # Fetch engagements
  result = JSON.parse(fetch_engagements(next_uri))
  engagements = result['engagements']
  # Append engagements to all_engagements array
  all_engagements += engagements

  # If next_page is present then continue with next_page provided by API
  # Else ends the loop
  next_uri = result['next_page'] ? URI(result['next_page']) : nil
end

# Prints out the amount of all engagements
puts all_engagements.count

File: rest_api/node/fetch_engagements.js
#!/usr/bin/env node

// Browsers have `fetch` built-in.
const fetch = require('node-fetch');

const token = 'YOUR BEARER TOKEN';
const siteId = 'YOUR SITE ID';

const perPage = 50;
const startDate = '2020-04-01T00:00:00Z';
const endDate = '2020-05-01T00:00:00Z';

console.log('Fetching engagements', {perPage, startDate, endDate});

function fetchEngagementsBatch(url) {
  return fetch(url, {
    method: 'get',
    json: true,
    headers: {
      Authorization: `Bearer ${token}`
    }
  }).then(res => res.json())
}

function fetchEngagements(url) {
  return fetchEngagementsBatch(url)
    .then(response => {
      if (response['next_page']) {
        // Fetch next page
        console.log('Fetching next page')
        return fetchEngagements(response['next_page'])
          .then(engagements => {
            // Merge current engagements and engagements from the next pages
            return Promise.resolve(response['engagements'].concat(engagements))
          });
      } else {
        console.log('No next_page, stopping.')
        return Promise.resolve(response['engagements']);
      }
    });
}

fetchEngagements(`https://api.salemove.com/engagements?per_page=${perPage}&start_date=${startDate}&end_date=${endDate}&site_ids[]=${siteId}`)
  .then(engagements => console.log('Fetched Engagements', engagements));
